applications/IgaApplication/custom_elements/shell_5p_hierarchic_element.cpp

```diff
@@ -0,0 +1,1307 @@
+            rValues = (stress_cau_cart[mGaussIntegrationThickness.num_GP_thickness-1][2] - stress_cau_cart_mid[2]) * thickness * thickness / 
+            (mGaussIntegrationThickness.zeta(mGaussIntegrationThickness.num_GP_thickness-1) * 6);
+        }
+        else if (rVariable == SHEAR_FORCE_1){
+            rValues = thickness * stress_cau_cart_mid[4];
+        }
+        else if (rVariable == SHEAR_FORCE_2){
+            rValues = thickness * stress_cau_cart_mid[3];
+        }
+        else{
+            KRATOS_WATCH("No results for desired variable available in Calculate of Shell5pHierarchicElement.")
+            rValues = 0.0;
+        }*/
+
+        KRATOS_CATCH("");
+    }
+
+    void Shell5pHierarchicElement::CalculateHessian(
+        Matrix& Hessian,
+        const Matrix& rDDN_DDe)
+    {
+        const SizeType number_of_points = GetGeometry().size();
+        const SizeType working_space_dimension = 3;
+        Hessian.resize(working_space_dimension, working_space_dimension);
+        Hessian = ZeroMatrix(working_space_dimension, working_space_dimension);
+
+        for (IndexType k = 0; k < number_of_points; k++)
+        {
+            const array_1d<double, 3> coords = GetGeometry()[k].Coordinates();
+
+            Hessian(0, 0) += rDDN_DDe(k, 0) * coords[0];
+            Hessian(0, 1) += rDDN_DDe(k, 2) * coords[0];
+            Hessian(0, 2) += rDDN_DDe(k, 1) * coords[0];
+
+            Hessian(1, 0) += rDDN_DDe(k, 0) * coords[1];
+            Hessian(1, 1) += rDDN_DDe(k, 2) * coords[1];
+            Hessian(1, 2) += rDDN_DDe(k, 1) * coords[1];
+
+            Hessian(2, 0) += rDDN_DDe(k, 0) * coords[2];
+            Hessian(2, 1) += rDDN_DDe(k, 2) * coords[2];
+            Hessian(2, 2) += rDDN_DDe(k, 1) * coords[2];
+        }
+    }
+
+    void Shell5pHierarchicElement::EquationIdVector(
+        EquationIdVectorType& rResult,
+        ProcessInfo& rCurrentProcessInfo
+    )
+    {
+        KRATOS_TRY
+
+        SizeType number_of_control_points = GetGeometry().size();
+
+        if (rResult.size() != 5 * number_of_control_points)
+            rResult.resize(5 * number_of_control_points, false);
+
+        for (IndexType i = 0; i < number_of_control_points; ++i) {
+            IndexType index = i * 5;
+            rResult[index]     = GetGeometry()[i].GetDof(DISPLACEMENT_X).EquationId();
+            rResult[index + 1] = GetGeometry()[i].GetDof(DISPLACEMENT_Y).EquationId();
+            rResult[index + 2] = GetGeometry()[i].GetDof(DISPLACEMENT_Z).EquationId();
+            // only ROTATION_X and ROTATION_Y used preliminarily, to avoid new declarations
+            // ROTATION_X = w_1 (first component of hierarchic shear difference vector)
+            // ROTATION_Y = w_2 (second component of hierarchic shear difference vector)
+            rResult[index + 3] = GetGeometry()[i].GetDof(ROTATION_X).EquationId();
+            rResult[index + 4] = GetGeometry()[i].GetDof(ROTATION_Y).EquationId();

```

Comment:
This would also benefit from #7385 to use a more appropriate naming.

Reply:
I agree!

Reply:
How would it change?

Reply:
I thought some more, consider also that by default we use `ROTATION` for rotational dofs

if you use some other then you will have to make some more adjustments to the solver which might not be trivial

=> even if it doesn't fit 100% your intention I would carefully consider using some different variable, this can have nasty consequences!

Reply:
We can quite easily use the variable with using the auxiliary_variables_list. However, having the additional third dimension is a bit a mess.

Reply:
That will only add the dofs
There are still more things like Scheme, ConvCrits etc which are optimized to work with Disp (and Rot)